addons/sale_coupon/models/sale_coupon_program.py

```diff
@@ -252,8 +251,8 @@ def _filter_programs_on_products(self, order):
             valid_products = program._get_valid_products(products)
             ordered_rule_products_qty = sum(products_qties[product] for product in valid_products)
             # Avoid program if 1 ordered foo on a program '1 foo, 1 free foo'
-            if program.promo_applicability == 'on_current_order' and \
-               program.reward_type == 'product' and program._get_valid_products(program.reward_product_id):
+            if valid_products and program.promo_applicability == 'on_current_order' and \
+               program.reward_type == 'product' and program.reward_product_id in valid_products:
```

Comment:
How can you be sure that `program.reward_product_id` is in `products` (and has been verified as valid for the program) ?
Furthermore, I doubt that a domain evaluation on 1 product is costly, wanna investigate that on the db :D.

Reply:
Well, if I I am not mistaken it is the entire point of this checking. If you look the comment just above, it states "Avoid program if 1 ordered foo on a program '1 foo, 1 free foo'" It looks logical and equivalent to me to check like this in this regards.
I understand that with a rule such as "Product A or Product B" with `valid_products` that would be Product A and the `reward_product_id` being product B. In such case, the behavior may act differently. Modifying the code like that look more faithful to the comment IMHO

Reply:
Like I said in the other comment, IMO behavioral change shouldn't happen in a perf PR :confused:.
As it's not tested, this old behavior maybe wasn't intended, but we still cannot be sure we understand the expected behavior developed some years ago (and the perf gain is quite low).